File: src/common/dataStructures/CircularListBaseIterator.cpp
#include "common/dataStructures/CircularListBaseIterator.h"

CircularListBaseIterator::CircularListBaseIterator(CircularListBase &list): currentElement(), currentList() {
    init(list);
}

CircularListBaseIterator::~CircularListBaseIterator() {
}

void
CircularListBaseIterator::init(CircularListBase &list) {
    currentList = &list;
    currentElement = currentList->last;
}

CircularListLink *
CircularListBaseIterator::next() {
    CircularListLink *response;

    if ( currentElement == nullptr ) {
        response = nullptr;
    } else {
        currentElement = currentElement->nextLink;
        response = currentElement;
    }

    if ( currentElement == currentList->last ) {
        currentElement = nullptr;
    }

    return response;
}

File: typescript/base/src/main/vsdk/toolkit/common/dataStructures/CircularListBaseIterator.ts
import { CircularListBase } from "./CircularListBase";
import { CircularListLink } from "./CircularListLink";

export class CircularListBaseIterator {
  private currentElement: CircularListLink | null;
  private currentList: CircularListBase | null;

  public constructor(list: CircularListBase) {
    this.currentElement = null;
    this.currentList = null;
    this.init(list);
  }

  public init(list: CircularListBase): void {
    this.currentList = list;
    this.currentElement = this.currentList.lastLink();
  }

  public next(): CircularListLink | null {
    let response: CircularListLink | null;

    if (this.currentElement === null) {
      response = null;
    }
    else {
      this.currentElement = this.currentElement.nextLink;
      response = this.currentElement;
    }

    if (this.currentList !== null && this.currentElement === this.currentList.lastLink()) {
      this.currentElement = null;
    }

    return response;
  }
}
